File: java/Samples/Biometrics/CS/LatentFingerprintSampleCS/LongActionDialog.cs
using System;
using System.Windows.Forms;

namespace Neurotec.Samples
{
	public partial class LongActionDialog : Form
	{
		#region Private constructor

		private LongActionDialog(string title)
		{
			InitializeComponent();
			lblTitle.Text = title;
		}

		#endregion

		#region Private fields

		private bool _canClose = false;
		private Delegate _target;
		private object[] _args;
		private Action _targetInvoke;
		private IAsyncResult _asyncResult;
		private object _targetResult;

		#endregion

		#region Public static methods

		public static object ShowDialog(IWin32Window owner, string title, Delegate target, params object [] args)
		{
			LongActionDialog dialog = new LongActionDialog(title)
			{
				_target = target,
				_args = args
			};
			dialog.ShowDialog();
			return dialog.CompleteAction();
		}

		public static object ShowDialog(IWin32Window owner, string title, Action target)
		{
			Delegate d = new MethodInvoker(target);
			return ShowDialog(owner, title, d, null);
		}

		#endregion

		#region Private methods

		private object CompleteAction()
		{
			_targetInvoke.EndInvoke(_asyncResult);
			return _targetResult;
		}

		private void StartAction()
		{
			_targetInvoke = () => _targetResult = _target.DynamicInvoke(_args);
			_targetInvoke.BeginInvoke(OnActionCompleted, null);
		}

		private void OnActionCompleted(IAsyncResult r)
		{
			BeginInvoke(new AsyncCallback(result =>
				{
					_asyncResult = result;
					_canClose = true;
					Close();
				}), r);
		}

		private void LongTaskFormFormClosing(object sender, FormClosingEventArgs e)
		{
			e.Cancel = !_canClose;
		}

		private void LongActionDialogShown(object sender, EventArgs e)
		{
			StartAction();
		}

		#endregion
	}
}

File: java/Samples/Biometrics/VB/AbisSampleVB/Common/LongActionDialog.vb
Imports System
Imports System.Windows.Forms

Partial Public Class LongActionDialog
	Inherits Form
#Region "Private constructor"

	Private Sub New(ByVal title As String)
		InitializeComponent()
		lblTitle.Text = title
	End Sub

#End Region

#Region "Private fields"

	Private _canClose As Boolean = False
	Private _target As System.Delegate
	Private _args() As Object
	Private _targetInvoke As Action
	Private _asyncResult As IAsyncResult
	Private _targetResult As Object

#End Region

#Region "Public static methods"

	Public Overloads Shared Function ShowDialog(ByVal owner As IWin32Window, ByVal title As String, ByVal target As System.Delegate, ByVal ParamArray args() As Object) As Object
		Dim dialog As New LongActionDialog(title) With {._target = target, ._args = args}
		dialog.ShowDialog()
		Return dialog.CompleteAction()
	End Function

	Public Overloads Shared Function ShowDialog(ByVal owner As IWin32Window, ByVal title As String, ByVal target As Action) As Object
		Return ShowDialog(owner, title, target, Nothing)
	End Function

#End Region

#Region "Private methods"

	Private Function CompleteAction() As Object
		_targetInvoke.EndInvoke(_asyncResult)
		Return _targetResult
	End Function

	Private Sub ActionDynamicInvoke()
		_targetResult = _target.DynamicInvoke(_args)
	End Sub

	Private Sub StartAction()
		_targetInvoke = AddressOf ActionDynamicInvoke
		_targetInvoke.BeginInvoke(AddressOf OnActionCompleted, Nothing)
	End Sub

	Private Sub OnActionCompleted(ByVal r As IAsyncResult)
		BeginInvoke(New AsyncCallback(AddressOf InvokeCompleted), r)
	End Sub
	Private Sub InvokeCompleted(ByVal result As Object)
		_asyncResult = result
		_canClose = True
		Close()
	End Sub

	Private Sub LongTaskFormFormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs) Handles Me.FormClosing
		e.Cancel = Not _canClose
	End Sub

	Private Sub LongActionDialogShown(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Shown
		StartAction()
	End Sub

#End Region
End Class
